mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -250,114 +240,142 @@ class KMeans private (
 
       // Find the sum and count of points mapping to each center
       val totalContribs = data.mapPartitions { points =>
-        val thisActiveCenters = bcActiveCenters.value
-        val runs = thisActiveCenters.length
-        val k = thisActiveCenters(0).length
-        val dims = thisActiveCenters(0)(0).vector.size
+        val thisCenters = bcCenters.value
+        val k = thisCenters.length
+        val dims = thisCenters(0).vector.size
+
+        val sums = Array.fill(k)(Vectors.zeros(dims))
+        val counts = Array.fill(k)(0L)
 
-        val sums = Array.fill(runs, k)(Vectors.zeros(dims))
-        val counts = Array.fill(runs, k)(0L)
+        val vectorOfPoints = new ArrayBuffer[Vector]()
+        val normOfPoints = new ArrayBuffer[Double]()
+        var numRows = 0
 
+        // Construct points matrix
         points.foreach { point =>
-          (0 until runs).foreach { i =>
-            val (bestCenter, cost) = KMeans.findClosest(thisActiveCenters(i), point)
-            costAccums(i) += cost
-            val sum = sums(i)(bestCenter)
-            axpy(1.0, point.vector, sum)
-            counts(i)(bestCenter) += 1
+          vectorOfPoints.append(point.vector)
+          normOfPoints.append(point.norm)
+          numRows += 1
+        }
+
+        val pointMatrix = if (isSparse) {
+          val coo = new ArrayBuffer[(Int, Int, Double)]()
+          vectorOfPoints.zipWithIndex.foreach { v =>
+            val sv = v._1.asInstanceOf[SparseVector]
+            sv.indices.indices.foreach { i =>
+              coo.append((v._2, sv.indices(i), sv.values(i)))
+            }
           }
+          SparseMatrix.fromCOO(numRows, dims, coo.toSeq)
+        } else {
+          new DenseMatrix(numRows, dims, vectorOfPoints.flatMap(_.toArray).toArray, true)
         }
 
-        val contribs = for (i <- 0 until runs; j <- 0 until k) yield {
-          ((i, j), (sums(i)(j), counts(i)(j)))
+        // Construct centers matrix
+        val vectorOfCenters = new ArrayBuffer[Double]()
+        val normOfCenters = new ArrayBuffer[Double]()
+        thisCenters.foreach { center =>
+          vectorOfCenters.appendAll(center.vector.toArray)
+          normOfCenters.append(center.norm)
+        }
+        val centerMatrix = new DenseMatrix(dims, k, vectorOfCenters.toArray)
+
+        val a2b2 = new ArrayBuffer[Double]()
+        val normOfPointsArray = normOfPoints.toArray
+        val normOfCentersArray = normOfCenters.toArray
+        for (i <- 0 until k; j <- 0 until numRows) {
+          a2b2.append(normOfPointsArray(j) * normOfPointsArray(j) +
+            normOfCentersArray(i) * normOfCentersArray(i))
+        }
+
+        val distanceMatrix = new DenseMatrix(numRows, k, a2b2.toArray)
+        gemm(-2.0, pointMatrix, centerMatrix, 1.0, distanceMatrix)
+
+        val vectorOfPointsArray = vectorOfPoints.toArray
+        distanceMatrix.transpose.toArray.grouped(k).toArray.map(_.zipWithIndex.min).zipWithIndex
```

Comment:
Given then complexity of this line, it might be helpful to unpack the tuple and give semantic meaning to its elements. Something like:

```scala
distanceMatrix.transpose.toArray.grouped(k).toArray.map(_.zipWithIndex.min).zipWithIndex
          .foreach { case ((cost, center), i) =>
            costAccums += cost
            val sum = sums(center)
            axpy(1.0, vectorOfPointsArray(i), sum)
            counts(center) += 1
          }
```